stablehlo/dialect/StablehloOps.cpp

```diff
@@ -2294,148 +2294,112 @@ LogicalResult AllGatherOp::verify() {
+                              int64_t feature_index, Value scale) {
+  auto operandType = operand.getType().cast<RankedTensorType>();
+  if (static_cast<int64_t>(feature_index) >= operandType.getRank())
+    return emitError(loc) << "expects feature_index to be smaller "
+                             "than the rank of operand type; got feature_index "
+                          << feature_index << ", and rank "
+                          << operandType.getRank() << ".";
+
+  if (static_cast<int64_t>(feature_index) < 0)
+    return emitError(loc) << "expects feature_index to be a "
+                          << "non-negative number, got "
+                          << static_cast<int64_t>(feature_index) << ".";
+
+  const int64_t featureCount = operandType.getShape()[feature_index];
   const int64_t scaleShape =
-      scale().getType().cast<RankedTensorType>().getShape()[0];
+      scale.getType().cast<RankedTensorType>().getShape()[0];
+  // As ODS enforces `scale`, `mean`, `variance`, `offset` are AllShapesMatch,
+  // this also infers that featureCount is aligned with them.
   if (scaleShape != featureCount)
-    return emitOpError() << "expects the size of scale factor to be "
-                            "same as the feature count,"
-                            " but the size of scale factor is "
-                         << scaleShape << " and the feature count is "
-                         << featureCount << ".";
+    return emitError(loc) << "expects the size of scale factor to be "
+                             "same as the feature count,"
+                             " but the size of scale factor is "
+                          << scaleShape << " and the feature count is "
+                          << featureCount << ".";
+
+  return success();
+}
+
+// Refer ODS for properties that are already enforced including shapes and
+// element types. This verifier includes additional checks.
+LogicalResult BatchNormGradOp::verify() {
+  if (failed(verifyBatchNorm(getLoc(), operand(), feature_index(), scale())))
+    return failure();
+  return success();
+}
+
+LogicalResult BatchNormGradOp::inferReturnTypeComponents(
+    MLIRContext* context, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  BatchNormGradOp::Adaptor adaptor(operands, attributes, regions);
+
+  RankedTensorType operandType =
+      adaptor.operand().getType().cast<RankedTensorType>();
+  inferredReturnShapes.emplace_back(operandType.getShape(),
+                                    operandType.getElementType());
 
+  const int64_t featureCount = operandType.getShape()[adaptor.feature_index()];
+  SmallVector<int64_t> featureShape(1, featureCount);
+  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
+  inferredReturnShapes.emplace_back(featureShape, operandType.getElementType());
```

Comment:
Can we say `{featureCount}` instead? (Here and below).

Reply:
Tried but not working. Though `SmallVector<int64_t> x = {1};` is valid expression, but `inferredReturnShapes.emplace_back({1}, operandType.getElementType());` is not.

Reply:
Got it. Then how about simply changing the definition of `featureShape` to `SmallVector<int64_t> featureShape{featureCount}`?

Reply:
Sure!